fuzz/fuzz_targets/archive.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![no_main]
+
+use libfuzzer_sys::fuzz_target;
+
+use std::fs::{File, OpenOptions};
+use std::io::{Cursor, Write, Read};
+use tar::{Archive, Builder, EntryType, Header};
+use tempfile::tempdir;
+use std::convert::TryInto;
+use std::str;
+
+fuzz_target!(|data: &[u8]| {
+    // Skip this iteration when data is not enough
+    if data.len() < 10 {
+        return;
+    }
+
+    // Create temp file and dir
+    let temp_dir = tempdir().unwrap();
+    let file_name = match str::from_utf8(&data[0..data.len().min(10)]) {
+        Ok(name) => name.to_string(),
+        Err(_) => "default_file_name".to_string(),
+    };
+    let dir_name = match str::from_utf8(&data[data.len().min(10)..data.len().min(20)]) {
+        Ok(name) => name.to_string(),
+        Err(_) => "default_dir_name".to_string(),
+    };
+    let temp_file_path = temp_dir.path().join(format!("{}_file.tar", file_name));
+
+    // Initialise builder and cursor
+    let mut builder = Builder::new(Vec::new());
+    let mut cursor = Cursor::new(data.to_vec());
+
+    // Choose an etnry type
+    let entry_type_byte = data[0];
+    let entry_type = match entry_type_byte % 5 {
+        0 => EntryType::Regular,
+        1 => EntryType::Directory,
+        2 => EntryType::Symlink,
+        3 => EntryType::hard_link(),
+        _ => EntryType::character_special(),
+    };
+
+    // Initilaise header
+    let mut header = Header::new_gnu();
+    let file_size = u64::from_le_bytes(
+        data.get(1..9)
```

Comment:
We're reading the same data for the filename, right? That's fine, but shouldn't we try to use independent sources for this? IOW let's just consistently read some bytes from data at each step instead of slicing it.

Reply:
Yes, I have fixed them.